### DISCUSSION TITLE
NURBS

### DISCUSSION THREAD

**Discussion by numerys** [JustinSDK/dotSCAD#18] [Ideas]
Nurbs support would be great:

https://en.wikipedia.org/wiki/Non-uniform_rational_B-spline

**Comment by JustinSDK** [comment]
[`bspline_curve`](https://openhome.cc/eGossip/OpenSCAD/lib3x-bspline_curve.html)?

**Reply by BrettRD** [reply]
The 2d case of bspline_curve, like [`bezier_surface`](https://openhome.cc/eGossip/OpenSCAD/BezierSurface.html)

**Reply by JustinSDK** [reply]
Encapsulate it into a function?

```openscad
use <bspline_curve.scad>;
use <function_grapher.scad>;

ctrl_pts = [
    [[0, 0, 20],  [60, 0, -35],   [90, 0, 60],    [200, 0, 5]],
    [[0, 50, 30], [100, 60, -25], [120, 50, 120], [200, 50, 5]],
    [[0, 100, 0], [60, 120, 35],  [90, 100, 60],  [200, 100, 45]],
    [[0, 150, 0], [60, 150, -35], [90, 180, 60],  [200, 150, 45]]
];

thickness = 2;
t_step = 0.05;
degrees = 2;

function bspline_surface(t_step, points, degrees, r_knots, c_knots, r_weights, c_weights) =
    let(
		row_pts = [for(i = [0:len(ctrl_pts) - 1]) 
		    bspline_curve(t_step, degrees, ctrl_pts[i], r_knots, r_weights)
		]
	)
	[
	    for(j = [0:len(row_pts[0]) - 1]) 
		bspline_curve(t_step, degrees,
			[for(i = [0:len(row_pts) - 1]) row_pts[i][j]],
			c_knots, c_weights
		) 
	];
	


function_grapher(bspline_surface(t_step, ctrl_pts, degrees), thickness);
```

![image](https://user-images.githubusercontent.com/3424741/120299081-42503800-c2fd-11eb-84f6-91ceb4639a12.png)

**Reply by numerys** [reply]
I see, is the "other way around" also possible? In Inkscape and other vector-graphic-software all points which the curve must go through are defined. The shape is changed flexible with tangential routes (angle and size). Maybe such a "Inkscape"-way is possible?

![Inkscape](https://user-images.githubusercontent.com/17107201/120311043-541ee480-c2d7-11eb-9dff-3da62155bd5e.png)

Such like [here](https://www.xnurbs.com/) would be even better.

**Reply by BrettRD** [reply]
The splines you've pictured are separate bezier curves (not nurbs) which start and stop at common points and have matching tangents.
The first control point in a bezier marks the terminating location, the next control point sets the tangent.
You can make multiple `bezier_surface`s continuous by making them share control mesh edges, and by making the control points nearest the edge co-linear between surfaces.  Using that, you can stitch an arbitrarily complex smooth curve out of patches of bezier surfaces.

xnurbs takes a very different approach. it uses a constraint solver to build one huge nurbs surface satisfying a list of constraints (your edges), then takes chunks out of the resulting nurbs surface between the constraining edges.  The major advantage to this approach is in their marketing material; it can make curves that are aesthetically consistent across the large patch by optimising a single parameter across the patch.

Something like Xnurbs can be built in OpenSCAD and it would use `bspline_surface` for final render.
It would need a numerical optimiser, a collection of error functions that can compare a surface to user-supplied constraints, and a Jacobian of the error-function in terms of the arguments to bspline_surface.
OpenSCAD 2021.01 included function-literals which should make solvers and optimisers much less painful to implement.

Don't wait around though, if you just want to join a sweeping curve to a few straight edges, a `bezier_surface` is a great tool for it, and `bspline_surface` has just eliminated the tedium of stitching spline control point tangents between patches.

**Comment by numerys** [comment]
Could something like [that](https://www.xnurbs.com/) be build on top of bezier_surface?

**Reply by numerys** [reply]
Of course in a scripted Openscad way. Non-clicky.
